### PR TITLE
Added support for BEYOND [ICML-2024] 

### PR DESCRIPTION
**Pull request by allenhzy**
# Description

This pull request adds the support of the BEYOND Detection method proposed in [1].

[1] Be Your Own Neighborhood: Detecting Adversarial Example by the Neighborhood Relations Built on Self-Supervised Learning. ICML. 2024[[Paper]](https://openreview.net/pdf?id=S4LqI6CcJ3)

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Unit Test

**Test Configuration**:
- OS: Ubuntu 18.04
- Python version: 3.9.19
- ART version or commit number: 1.18.1
- Pytorch: 2.4.0
- cudnn version: 90100

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] My changes have been tested using both CPU and GPU devices

### CODE REVIEW

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @allenhzy Thank you very much for your pull request! Could you please take a look at my review comments and add the proposed updates?

File: art/defences/detector/evasion/beyond_detector.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the abstract base class for all evasion detectors.
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals, annotations
+
+import abc
```
[the remark this answers is not in the document]
Comment (allenhzy) [reply art/defences/detector/evasion/beyond_detector.py]: I've revised all codes that need to be revised

File: tests/defences/detector/evasion/test_beyond_detector.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    return model
+
+@pytest.mark.only_with_platform("pytorch")
+def test_beyond_detector(art_warning, get_cifar10, get_ssl_model):
+    try:
+        # Load CIFAR10 data
+        (x_train, y_train), (x_test, y_test), min_, max_ = get_cifar10
+
+        # Load models
+        # Download pretrained weights from https://drive.google.com/drive/folders/1ieEdd7hOj2CIl1FQfu4-3RGZmEj-mesi?usp=sharing
```
[the remark this answers is not in the document]
Comment (allenhzy) [reply tests/defences/detector/evasion/test_beyond_detector.py:110]: > The pre-trained model has over 100 M

**Review by beat-buesser — COMMENTED** [review]
Hi @allenhzy Thank you very much for your updates! Could you please take a look above at my two last change requests and the alerts from CodeQL about unused variables?

File: art/defences/detector/evasion/beyond_detector.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+This module implements the BEYOND detector for adversarial examples detection.
+| Paper link: https://openreview.net/pdf?id=S4LqI6CcJ3
+"""
+import numpy as np
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+
+
+logger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (allenhzy) [reply art/defences/detector/evasion/beyond_detector.py]: I have revised the related codes.

**Comment by allenhzy** [comment]
Dear Beat,

I have revised the codes.
Please check it.

Best,
Zhiyuan HE
发件人: Beat Buesser ***@***.***>
日期: 星期一, 2024年12月9日 下午7:29
收件人: Trusted-AI/adversarial-robustness-toolbox ***@***.***>
抄送: hzy ***@***.***>, Mention ***@***.***>
主题: Re: [Trusted-AI/adversarial-robustness-toolbox] Added support for BEYOND [ICML-2024] (PR #2489)

@beat-buesser commented on this pull request.

Hi @allenhzy<https://github.com/allenhzy> Thank you very much for your updates! Could you please take a look above at my two last change requests and the alerts from CodeQL about unused variables?

________________________________

In art/defences/detector/evasion/beyond_detector.py<https://github.com/Trusted-AI/adversarial-robustness-toolbox/pull/2489#discussion_r1875734678>:

> +"""

+This module implements the BEYOND detector for adversarial examples detection.

+| Paper link: https://openreview.net/pdf?id=S4LqI6CcJ3

+"""
⬇️ Suggested change

-"""

-This module implements the BEYOND detector for adversarial examples detection.

-| Paper link: https://openreview.net/pdf?id=S4LqI6CcJ3

-"""

+"""

+This module implements the BEYOND detector for adversarial examples detection.

+

+| Paper link: https://openreview.net/pdf?id=S4LqI6CcJ3

+"""

________________________________

In art/defences/detector/evasion/beyond_detector.py<https://github.com/Trusted-AI/adversarial-robustness-toolbox/pull/2489#discussion_r1875736394>:

> +# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE

+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE

+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,

+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE

+# SOFTWARE.

+"""

+This module implements the BEYOND detector for adversarial examples detection.

+| Paper link: https://openreview.net/pdf?id=S4LqI6CcJ3

+"""

+import numpy as np

+from typing import TYPE_CHECKING

+if TYPE_CHECKING:

+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE

+

+

+logger = logging.getLogger(__name__)

This line seems to cause the unittest errors because of the missing import of logging. Because the attack here is not actively using logging I suggest to remove this line.
⬇️ Suggested change

-logger = logging.getLogger(__name__)

File: art/defences/detector/evasion/beyond_detector.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    
+    | Paper link: https://openreview.net/pdf?id=S4LqI6CcJ3
+    """
+    
+    defence_params = ["target_model", "ssl_model", "augmentations", "aug_num", "alpha", "K", "percentile"]
+
+    def __init__(self,
+        target_model: "CLASSIFIER_NEURALNETWORK_TYPE",
+        ssl_model: "CLASSIFIER_NEURALNETWORK_TYPE",
+        augmentations: Optional[Callable] = None,
```
Comment (beat-buesser) [inline art/defences/detector/evasion/beyond_detector.py:44]:
```suggestion
        augmentations: Callable | None,
```

File: art/defences/detector/evasion/beyond_detector.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the BEYOND detector for adversarial examples detection.
+
+| Paper link: https://openreview.net/pdf?id=S4LqI6CcJ3
+"""
+import numpy as np
```
Comment (beat-buesser) [inline art/defences/detector/evasion/beyond_detector.py:23]:
```suggestion
from __future__ import annotations

import numpy as np
```

File: tests/defences/detector/evasion/test_beyond_detector.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
```
Comment (beat-buesser) [inline tests/defences/detector/evasion/test_beyond_detector.py:20]:
```suggestion
```

**Comment by beat-buesser** [comment]
Hi @allenhzy I have merge this PR into the release branch, but I had to make some changes to remove errors that prevented the detector from running at all. Could you please take a look at the changes [here](https://github.com/Trusted-AI/adversarial-robustness-toolbox/commit/5ec99dc29dd501dda478fe39c5cfa3ffd7d87e9c) to see if they are correct and let me know?

**Pull request closed without merging**